internal/v4/list.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2014 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package v4 // import "gopkg.in/juju/charmstore.v5-unstable/internal/v4"
+
+import (
+	"net/http"
+	"sync/atomic"
+
+	"github.com/juju/utils/parallel"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/router"
+)
+
+// GET list[?filter=value…][&include=meta][&sort=field[+dir]]
+// https://github.com/juju/charmstore/blob/v4/docs/API.md#get-list
+func (h *ReqHandler) serveList(_ http.Header, req *http.Request) (interface{}, error) {
+	lp, err := parseListParams(req)
+	if err != nil {
+		return "", err
+	}
+	return h.doList(lp, req)
+}
+
+// doList performs a listing specified by ListParams. If lp
+// specifies that additional metadata needs to be added to the results,
+// then it is added. It also filters the result based on ACLs.
+func (h *ReqHandler) doList(lp charmstore.ListParams, req *http.Request) (interface{}, error) {
+	// perform query
+	results, err := h.Store.List(lp)
+	if err != nil {
+		return nil, errgo.Notef(err, "error listing charms and bundles")
+	}
+	filteredACLResults := make([]*router.ResolvedURL, 0)
+	for _, result := range results.Results {
+		if err = h.AuthorizeEntity(result, req); err == nil {
+			filteredACLResults = append(filteredACLResults, result)
+		}
+	}
+	results.Results = filteredACLResults
+	return h.addMetaData(results, lp.Include, req)
+}
+
+//addMetada adds the requested meta data with the include list.
```

Comment:
Most of this code is copy and pasted from ReqHandler.doSearch. Let's please use the same code for both as they're both doing almost exactly the same thing. (I think that will be easier when both are using EntityResult).